File: internal/awsx/insights.go
package awsx

import (
	"context"
	"errors"
	"fmt"
	"sort"
	"sync"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs"
	logtypes "github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs/types"

	"github.com/jhyeok1023/skills-dashboard/internal/domain"
)

// InsightsMaxRows is the ceiling CloudWatch puts on a Logs Insights result set.
// A query that returns exactly this many rows was probably cut off, and the
// payload says so rather than presenting a partial answer as complete.
const InsightsMaxRows = 10000

// QueryResult is one completed Logs Insights query.
type QueryResult struct {
	ID   string
	Rows []map[string]string

	// BytesScanned is what the query cost. Insights bills per byte scanned, so
	// this is surfaced in the UI: the operator can see what a refresh costs
	// instead of discovering it on a bill.
	BytesScanned   float64
	RecordsMatched float64
	Truncated      bool
	Elapsed        time.Duration
}

// InsightsRunner executes Logs Insights queries with the guard rails the
// reference implementation lacked: a bounded number in flight, a deadline on
// each one, and a StopQuery on the way out.
type InsightsRunner struct {
	API LogsAPI

	// Concurrency bounds queries in flight. CloudWatch allows roughly thirty
	// concurrent queries per account; a page here issues five or six, and
	// staying well under leaves headroom for anything else using the key.
	Concurrency int
	// Timeout bounds a single query.
	Timeout time.Duration
	// PollInterval is how often GetQueryResults is called while a query runs.
	PollInterval time.Duration

	once sync.Once
	sem  chan struct{}
}

func (r *InsightsRunner) init() {
	r.once.Do(func() {
		n := r.Concurrency
		if n <= 0 {
			n = 6
		}
		r.sem = make(chan struct{}, n)
		if r.Timeout <= 0 {
			r.Timeout = 45 * time.Second
		}
		if r.PollInterval <= 0 {
			r.PollInterval = 250 * time.Millisecond
		}
	})
}

// ErrNoLogGroup is returned when a query is requested for a log group that has
// not been configured, so the caller can render an explanatory panel instead of
// an error.
var ErrNoLogGroup = errors.New("no log group configured")

// Run executes one query over the window and waits for it to finish.
func (r *InsightsRunner) Run(ctx context.Context, logGroup string, w domain.Window, q domain.Query) (QueryResult, error) {
	r.init()
	if logGroup == "" {
		return QueryResult{}, ErrNoLogGroup
	}

	select {
	case r.sem <- struct{}{}:
		defer func() { <-r.sem }()
	case <-ctx.Done():
		return QueryResult{}, ctx.Err()
	}

	ctx, cancel := context.WithTimeout(ctx, r.Timeout)
	defer cancel()

	started := time.Now()
	in := &cloudwatchlogs.StartQueryInput{
		LogGroupNames: []string{logGroup},
		StartTime:     aws.Int64(w.Start.Unix()),
		EndTime:       aws.Int64(w.End.Unix()),
		QueryString:   aws.String(q.Text),
	}
	if q.Limit > 0 {
		in.Limit = aws.Int32(int32(q.Limit))
	}

	out, err := r.API.StartQuery(ctx, in)
	if err != nil {
		return QueryResult{}, fmt.Errorf("StartQuery %s: %w", q.ID, err)
	}
	queryID := aws.ToString(out.QueryId)
	if queryID == "" {
		return QueryResult{}, fmt.Errorf("StartQuery %s returned no query id", q.ID)
	}

	res, err := r.poll(ctx, queryID, q)
	if err != nil {
		// The query is still running on CloudWatch's side and still holding a
		// concurrency slot. Releasing it needs a live context, so a fresh one
		// is used rather than the cancelled one we are unwinding from.
		r.stop(queryID)
		return QueryResult{}, err
	}
	res.Elapsed = time.Since(started)
	return res, nil
}

func (r *InsightsRunner) poll(ctx context.Context, queryID string, q domain.Query) (QueryResult, error) {
	ticker := time.NewTicker(r.PollInterval)
	defer ticker.Stop()

	for {
		out, err := r.API.GetQueryResults(ctx, &cloudwatchlogs.GetQueryResultsInput{
			QueryId: aws.String(queryID),
		})
		if err != nil {
			return QueryResult{}, fmt.Errorf("GetQueryResults %s: %w", q.ID, err)
		}

		switch out.Status {
		case logtypes.QueryStatusComplete:
			return buildResult(q, out), nil
		case logtypes.QueryStatusFailed:
			return QueryResult{}, fmt.Errorf("query %s failed", q.ID)
		case logtypes.QueryStatusCancelled:
			return QueryResult{}, fmt.Errorf("query %s was cancelled", q.ID)
		case logtypes.QueryStatusTimeout:
			return QueryResult{}, fmt.Errorf("query %s timed out in CloudWatch", q.ID)
		}

		select {
		case <-ctx.Done():
			return QueryResult{}, fmt.Errorf("query %s: %w", q.ID, ctx.Err())
		case <-ticker.C:
		}
	}
}

func buildResult(q domain.Query, out *cloudwatchlogs.GetQueryResultsOutput) QueryResult {
	res := QueryResult{ID: q.ID, Rows: make([]map[string]string, 0, len(out.Results))}
	for _, row := range out.Results {
		m := make(map[string]string, len(row))
		for _, f := range row {
			name := aws.ToString(f.Field)
			if name == "@ptr" {
				continue
			}
			m[name] = aws.ToString(f.Value)
		}
		res.Rows = append(res.Rows, m)
	}
	if s := out.Statistics; s != nil {
		res.BytesScanned = s.BytesScanned
		res.RecordsMatched = s.RecordsMatched
	}

	limit := q.Limit
	if limit <= 0 || limit > InsightsMaxRows {
		limit = InsightsMaxRows
	}
	res.Truncated = len(res.Rows) >= limit
	return res
}

// stop releases the CloudWatch-side query slot. It runs on its own short-lived
// context because the caller's has usually just been cancelled, and it ignores
// the error: the query may well have finished on its own, and there is nothing
// useful to do about a failure to cancel something already gone.
func (r *InsightsRunner) stop(queryID string) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	_, _ = r.API.StopQuery(ctx, &cloudwatchlogs.StopQueryInput{QueryId: aws.String(queryID)})
}

// RunAll executes every query concurrently, subject to the runner's limit, and
// returns the results keyed by query id.
//
// One query failing does not sink the rest: its error is recorded against its
// id so the panel it feeds can say what went wrong while its neighbours still
// render.
func (r *InsightsRunner) RunAll(ctx context.Context, logGroup string, w domain.Window, qs []domain.Query) (map[string]QueryResult, map[string]error) {
	r.init()
	results := make(map[string]QueryResult, len(qs))
	errs := map[string]error{}

	var mu sync.Mutex
	var wg sync.WaitGroup
	for _, q := range qs {
		wg.Add(1)
		go func(q domain.Query) {
			defer wg.Done()
			res, err := r.Run(ctx, logGroup, w, q)
			mu.Lock()
			defer mu.Unlock()
			if err != nil {
				errs[q.ID] = err
				return
			}
			results[q.ID] = res
		}(q)
	}
	wg.Wait()
	return results, errs
}

// TotalBytesScanned sums what a set of results cost.
func TotalBytesScanned(results map[string]QueryResult) float64 {
	var total float64
	for _, r := range results {
		total += r.BytesScanned
	}
	return total
}

// SortedRows returns a result's rows ordered by the named field, which keeps a
// table from reshuffling between refreshes when CloudWatch returns ties in a
// different order.
func SortedRows(rows []map[string]string, field string) []map[string]string {
	out := append([]map[string]string(nil), rows...)
	sort.SliceStable(out, func(i, j int) bool { return out[i][field] < out[j][field] })
	return out
}

File: server/src/aws/insights.ts
// Logs Insights 쿼리를 돌린다. internal/awsx/insights.go 의 이식이다.
//
// 이전 구현에 없던 난간을 단다: 동시에 나는 쿼리 수의 상한, 쿼리마다의 시한,
// 그리고 나가는 길의 StopQuery.

import {
	GetQueryResultsCommand,
	StartQueryCommand,
	StopQueryCommand,
	type CloudWatchLogsClient
} from '@aws-sdk/client-cloudwatch-logs';

import type { Query } from '../domain/query.ts';
import type { Window } from '../domain/window.ts';
import { sendOptions } from './client.ts';

/**
 * CloudWatch 가 Logs Insights 결과 집합에 씌우는 천장. 정확히 이만큼의 행을
 * 돌려준 쿼리는 잘렸을 가능성이 크고, payload 는 부분적인 답을 완전한 것처럼
 * 내놓는 대신 그렇다고 말한다.
 */
export const insightsMaxRows = 10_000;

/** 완료된 Insights 쿼리 하나. */
export interface QueryResult {
	id: string;
	rows: Record<string, string>[];

	/**
	 * 이 쿼리가 든 비용. Insights 는 스캔한 바이트로 과금하므로 UI 에 드러낸다.
	 * 운영자가 새로고침 한 번의 값을 청구서에서 발견하는 대신 화면에서 본다.
	 */
	bytesScanned: number;
	recordsMatched: number;
	truncated: boolean;
	elapsedMs: number;
}

/**
 * 설정되지 않은 로그 그룹에 쿼리를 요청했다는 뜻. 호출자가 오류 대신 설명하는
 * 패널을 그릴 수 있도록 따로 구분한다.
 */
export class NoLogGroupError extends Error {
	constructor() {
		super('no log group configured');
		this.name = 'NoLogGroupError';
	}
}

/**
 * 동시 실행 수를 묶는 세마포어.
 *
 * Go 는 버퍼 채널로 같은 일을 했다. 기다리는 동안 호출자가 사라지면 자리를
 * 기다리던 것도 함께 접어야 한다 — 안 그러면 아무도 읽지 않을 답을 위해 유료
 * 스캔이 시작된다.
 */
class Semaphore {
	private free: number;
	private readonly waiting: (() => void)[] = [];

	constructor(n: number) {
		this.free = n;
	}

	async acquire(signal?: AbortSignal): Promise<void> {
		if (signal?.aborted === true) throw signal.reason ?? new Error('aborted');
		if (this.free > 0) {
			this.free--;
			return;
		}
		await new Promise<void>((resolve, reject) => {
			const onAbort = () => {
				const i = this.waiting.indexOf(grant);
				if (i >= 0) this.waiting.splice(i, 1);
				reject(signal?.reason ?? new Error('aborted'));
			};
			const grant = () => {
				signal?.removeEventListener('abort', onAbort);
				resolve();
			};
			this.waiting.push(grant);
			signal?.addEventListener('abort', onAbort, { once: true });
		});
		this.free--;
	}

	release(): void {
		this.free++;
		const next = this.waiting.shift();
		if (next !== undefined) next();
	}
}

export interface InsightsRunnerOptions {
	/**
	 * 동시에 나는 쿼리 수의 상한. CloudWatch 는 계정당 서른 개쯤을 허용하고
	 * 여기서 한 페이지는 대여섯 개를 낸다. 넉넉히 아래에 두면 같은 키를 쓰는
	 * 다른 것에도 여유가 남는다.
	 */
	concurrency?: number;
	/** 쿼리 하나의 시한(ms). */
	timeoutMs?: number;
	/** 쿼리가 도는 동안 GetQueryResults 를 부르는 간격(ms). */
	pollIntervalMs?: number;
}

export class InsightsRunner {
	readonly api: CloudWatchLogsClient;
	private readonly sem: Semaphore;
	private readonly timeoutMs: number;
	private readonly pollIntervalMs: number;

	constructor(api: CloudWatchLogsClient, options: InsightsRunnerOptions = {}) {
		this.api = api;
		const n = options.concurrency !== undefined && options.concurrency > 0 ? options.concurrency : 6;
		this.sem = new Semaphore(n);
		this.timeoutMs =
			options.timeoutMs !== undefined && options.timeoutMs > 0 ? options.timeoutMs : 45_000;
		this.pollIntervalMs =
			options.pollIntervalMs !== undefined && options.pollIntervalMs > 0
				? options.pollIntervalMs
				: 250;
	}

	/** run 은 창 위에서 쿼리 하나를 돌리고 끝날 때까지 기다린다. */
	async run(logGroup: string, w: Window, q: Query, signal?: AbortSignal): Promise<QueryResult> {
		if (logGroup === '') throw new NoLogGroupError();

		await this.sem.acquire(signal);
		try {
			return await this.runHeld(logGroup, w, q, signal);
		} finally {
			this.sem.release();
		}
	}

	private async runHeld(
		logGroup: string,
		w: Window,
		q: Query,
		signal?: AbortSignal
	): Promise<QueryResult> {
		const deadline = AbortSignal.timeout(this.timeoutMs);
		const bounded = signal === undefined ? deadline : AbortSignal.any([signal, deadline]);

		const started = Date.now();
		let queryId = '';
		try {
			const out = await this.api.send(
				new StartQueryCommand({
					logGroupNames: [logGroup],
					startTime: Math.trunc(w.start / 1000),
					endTime: Math.trunc(w.end / 1000),
					queryString: q.text,
					...(q.limit > 0 ? { limit: q.limit } : {})
				}),
				sendOptions(bounded)
			);
			queryId = out.queryId ?? '';
		} catch (err) {
			throw new Error(`StartQuery ${q.id}: ${message(err)}`);
		}
		if (queryId === '') throw new Error(`StartQuery ${q.id} returned no query id`);

		try {
			const res = await this.poll(queryId, q, bounded);
			res.elapsedMs = Date.now() - started;
			return res;
		} catch (err) {
			// 쿼리는 CloudWatch 쪽에서 아직 돌고 있고 동시 실행 자리를 붙들고
			// 있다. 그것을 놓아주려면 살아 있는 신호가 필요하므로, 지금 풀려
			// 나가는 중인 것 대신 새 시한을 쓴다.
			this.stop(queryId);
			throw err;
		}
	}

	private async poll(queryId: string, q: Query, signal: AbortSignal): Promise<QueryResult> {
		for (;;) {
			let out;
			try {
				out = await this.api.send(new GetQueryResultsCommand({ queryId }), sendOptions(signal));
			} catch (err) {
				throw new Error(`GetQueryResults ${q.id}: ${message(err)}`);
			}

			switch (out.status) {
				case 'Complete':
					return buildResult(q, out.results ?? [], out.statistics);
				case 'Failed':
					throw new Error(`query ${q.id} failed`);
				case 'Cancelled':
					throw new Error(`query ${q.id} was cancelled`);
				case 'Timeout':
					throw new Error(`query ${q.id} timed out in CloudWatch`);
				default:
					break;
			}

			await wait(this.pollIntervalMs, signal, q.id);
		}
	}

	/**
	 * stop 은 CloudWatch 쪽 쿼리 자리를 놓아준다. 짧은 제 시한으로 돌고 오류는
	 * 무시한다 — 쿼리가 스스로 끝났을 수도 있고, 이미 사라진 것을 취소하지 못한
	 * 데 대해 할 수 있는 유용한 일이 없다.
	 */
	private stop(queryId: string): void {
		void this.api
			.send(new StopQueryCommand({ queryId }), sendOptions(AbortSignal.timeout(5_000)))
			.catch(() => {});
	}

	/**
	 * runAll 은 러너의 상한 안에서 모든 쿼리를 동시에 돌리고 결과를 쿼리 id 로
	 * 묶어 돌려준다.
	 *
	 * 쿼리 하나가 실패해도 나머지가 가라앉지 않는다. 오류는 그 id 앞으로
	 * 기록되므로, 그것이 먹이는 패널은 무엇이 잘못됐는지 말하면서 이웃 패널은
	 * 그대로 그려진다.
	 */
	async runAll(
		logGroup: string,
		w: Window,
		queries: Query[],
		signal?: AbortSignal
	): Promise<{ results: Map<string, QueryResult>; errors: Map<string, Error> }> {
		const results = new Map<string, QueryResult>();
		const errors = new Map<string, Error>();

		await Promise.all(
			queries.map(async (q) => {
				try {
					results.set(q.id, await this.run(logGroup, w, q, signal));
				} catch (err) {
					errors.set(q.id, err instanceof Error ? err : new Error(String(err)));
				}
			})
		);
		return { results, errors };
	}
}

function buildResult(
	q: Query,
	rows: { field?: string | undefined; value?: string | undefined }[][],
	statistics?: { bytesScanned?: number | undefined; recordsMatched?: number | undefined }
): QueryResult {
	const out: QueryResult = {
		id: q.id,
		rows: [],
		bytesScanned: statistics?.bytesScanned ?? 0,
		recordsMatched: statistics?.recordsMatched ?? 0,
		truncated: false,
		elapsedMs: 0
	};

	for (const row of rows) {
		const m: Record<string, string> = {};
		for (const f of row) {
			const name = f.field ?? '';
			if (name === '@ptr') continue;
			m[name] = f.value ?? '';
		}
		out.rows.push(m);
	}

	const limit = q.limit <= 0 || q.limit > insightsMaxRows ? insightsMaxRows : q.limit;
	out.truncated = out.rows.length >= limit;
	return out;
}

function wait(ms: number, signal: AbortSignal, queryID: string): Promise<void> {
	return new Promise((resolve, reject) => {
		const onAbort = () => {
			clearTimeout(timer);
			reject(new Error(`query ${queryID}: ${message(signal.reason)}`));
		};
		const timer = setTimeout(() => {
			signal.removeEventListener('abort', onAbort);
			resolve();
		}, ms);
		if (signal.aborted) {
			onAbort();
			return;
		}
		signal.addEventListener('abort', onAbort, { once: true });
	});
}

function message(err: unknown): string {
	return err instanceof Error ? err.message : String(err);
}

/** totalBytesScanned 는 결과 묶음이 든 비용을 더한다. */
export function totalBytesScanned(results: Map<string, QueryResult>): number {
	let total = 0;
	for (const r of results.values()) total += r.bytesScanned;
	return total;
}

/**
 * sortedRows 는 결과의 행을 이름 붙인 필드로 정렬한다. CloudWatch 가 동점을
 * 다른 순서로 돌려줘도 표가 새로고침마다 뒤섞이지 않게 한다.
 */
export function sortedRows(
	rows: Record<string, string>[],
	field: string
): Record<string, string>[] {
	return [...rows].sort((a, b) => {
		const x = a[field] ?? '';
		const y = b[field] ?? '';
		return x < y ? -1 : x > y ? 1 : 0;
	});
}
